prow/config/config.go

```diff
@@ -264,6 +342,70 @@ func Load(path string) (*Config, error) {
 	return nc, nil
 }
 
+// mergeConfig merges two JobConfig together
+// It will try to merge:
+//	- Presubmits
+//	- Postsubmits
+// 	- Periodics
+//	- PodPresets
+func (c *Config) mergeJobConfig(jc JobConfig) error {
+	c.Presets = append(c.Presets, jc.Presets...)
+
+	c.Periodics = append(c.Periodics, jc.Periodics...)
+
+	for repo, jobs := range jc.Presubmits {
+		if _, ok := c.Presubmits[repo]; ok {
+			c.Presubmits[repo] = append(c.Presubmits[repo], jobs...)
+		}
+	}
+
+	for repo, jobs := range jc.Postsubmits {
+		if _, ok := c.Postsubmits[repo]; ok {
+			c.Postsubmits[repo] = append(c.Postsubmits[repo], jobs...)
+		}
+	}
+
+	// validate no duplicated presets
+	validLabels := map[string]string{}
+	for _, preset := range c.Presets {
+		for label, val := range preset.Labels {
+			if _, ok := validLabels[label]; ok {
+				return fmt.Errorf("duplicated preset label : %s", label)
+			}
+			validLabels[label] = val
+		}
+	}
+
+	// validate no duplicated periodics
+	validPeriodics := map[string]bool{}
+	for _, p := range c.AllPeriodics() {
+		if _, ok := validPeriodics[p.Name]; ok {
+			return fmt.Errorf("duplicated periodic job : %s", p.Name)
+		}
+		validPeriodics[p.Name] = true
+	}
+
+	// validate no duplicated presubmits
+	validPresubmits := map[string]bool{}
+	for _, p := range c.AllPresubmits(nil) {
+		if _, ok := validPresubmits[p.Name]; ok {
+			return fmt.Errorf("duplicated presubmit job : %s", p.Name)
```

Comment:
Actually this won't allow multiple jobs with the same name (even within a file) so this is too strict for the branch-sharded jobs that we have in k/k.

It could be reasonable to instead enforce that all presubmits for a repo are within one file (same for postsubmits)?

Reply:
ah good point there! I'll make the fix

Reply:
added more logics and tests, basically we allow multiple presubmits/postsubmits from a same job config file.